### PR TITLE
Added benchmarks for link generation

### PR DESCRIPTION
**Pull request by kichalla**

### CODE REVIEW

File: benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/GithubRoutesBenchmark.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using BenchmarkDotNet.Attributes;
+using Microsoft.AspNetCore.Routing.Matching;
+using Microsoft.AspNetCore.Routing.Tree;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Routing.LinkGeneration
+{
+    public class GithubRoutesBenchmark : LinkGenerationBenchmarkBase
```
Comment (kichalla) [inline benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/GithubRoutesBenchmark.cs]: image

File: benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/SingleRouteWithNoParametersBenchmark.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using BenchmarkDotNet.Attributes;
+using Microsoft.AspNetCore.Routing.Tree;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Routing.LinkGeneration
+{
+    public class SingleRouteWithNoParametersBenchmark : LinkGenerationBenchmarkBase
```
Comment (kichalla) [inline benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/SingleRouteWithNoParametersBenchmark.cs]: image

File: benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/LinkGenerationBenchmarkBase.cs [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+                // Increment the action count within a controller template
+                visitedTemplateInfo.Item2++;
+
+                var requiredValues = new
+                {
+                    area = (string)null,
+                    controller = $"Controller{visitedTemplateInfo.Item1}",
+                    action = $"Action{visitedTemplateInfo.Item2}",
+                    page = (string)null
+                };
```
Comment (rynowak) [inline benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/LinkGenerationBenchmarkBase.cs]: I think this is a pretty good approach, and somewhat realistic as long as your considering APIs (not web sites). Nice 👍

File: benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/RequestContext.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Routing.LinkGeneration
+{
+    public class RequestContext
```
Comment (rynowak) [inline benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/RequestContext.cs]: use a valuetuple?

File: benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/SingleRouteRouteValuesBasedEndpointFinderBenchmark.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using BenchmarkDotNet.Attributes;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Routing.LinkGeneration
+{
+    public class SingleRouteRouteValuesBasedEndpointFinderBenchmark : LinkGenerationBenchmarkBase
```
Comment (rynowak) [inline benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/SingleRouteRouteValuesBasedEndpointFinderBenchmark.cs]: BTW if you want, it's possible to combine all of these cases into a single benchmark class.

File: benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/SingleRouteRouteValuesBasedEndpointFinderBenchmark.cs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            _baseFinder = new TestEndpointFinder(endpoint);
+
+            _requestContext = CreateCurrentRequestContext();
+        }
+
+        [Benchmark(Baseline = true)]
+        public void Baseline()
+        {
+            var actual = _baseFinder.FindEndpoints(address: 0);
+            GC.KeepAlive(actual);
```
Comment (JamesNK) [inline benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/SingleRouteRouteValuesBasedEndpointFinderBenchmark.cs]: What is the purpose of the KeepAlive? Either remove or add a comment.

Repeat for other benchmarks in this file.

File: benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/SingleRouteWithConstraintsBenchmark.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            _requestContext = CreateCurrentRequestContext();
+        }
+
+        [Benchmark(Baseline = true)]
+        public void TreeRouter()
+        {
+            var virtualPathData = _treeRouter.GetVirtualPath(new VirtualPathContext(
+                _requestContext.HttpContext,
+                ambientValues: _requestContext.AmbientValues,
+                values: new RouteValueDictionary(
```
Comment (JamesNK) [inline benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/SingleRouteWithConstraintsBenchmark.cs]: IMO initialize the RVD once in Setup and reuse it. It reduces some of the allocation noise in the results.
Could do this for all benchmarks.

File: benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/LinkGenerationBenchmarkBase.cs [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+
+            return services.BuildServiceProvider();
+        }
+
+        protected void AssertUrl(string expectedUrl, string actualUrl)
+        {
+            AssertUrl(expectedUrl, actualUrl, StringComparison.Ordinal);
+        }
+
+        protected void AssertUrl(string expectedUrl, string actualUrl, StringComparison stringComparison)
```
Comment (JamesNK) [inline benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/LinkGenerationBenchmarkBase.cs]: I like it. It is easy for micro-benchmarks to start doing the wrong thing and no one notices

File: benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/LinkGenerationBenchmarkBase.cs [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+            var requiredValues = routeValuesAddressMetadata?.RequiredValues ?? new RouteValueDictionary();
+
+            AddOutboundRouteEntry(
+                treeRouteBuilder,
+                endpoint.RoutePattern.RawText,
+                defaults: endpoint.RoutePattern.Defaults,
+                requiredValues: requiredValues);
+        }
+
+        protected MatcherEndpoint[] GetGithubEndpoints()
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/LinkGenerationBenchmarkBase.cs]: Add a comment saying where the GitHub endpoints came from and how to generate them.

**Review by kichalla — COMMENTED** [review]
🆙 📅 Does not have the entire feedback that you gave (for example, doesn't have the feedback related to generating links to all endpoints in Github) but could you take a look at endpoint generation and let me know if it looks ok

File: benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/SingleRouteWithConstraintsBenchmark.cs (hunk 7 above)
Comment (rynowak) [reply benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/SingleRouteWithConstraintsBenchmark.cs]: I agree with Kiran here. Typically a user is passing a new RVD into each callsite.

**Review by rynowak — APPROVED** [review]

File: benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/LinkGenerationGithubBenchmark.generated.cs [hunk 10]
```diff
@@ -0,0 +1,1519 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing.Matching;
+
+namespace Microsoft.AspNetCore.Routing.LinkGeneration
+{
+    // This code was generated by the Swaggatherer
+    public partial class LinkGenerationGithubBenchmark : EndpointRoutingBenchmarkBase
```
[the remark this answers is not in the document]
Comment (rynowak) [reply benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/LinkGenerationGithubBenchmark.generated.cs]: I think this is fine for now. I'll not ask you to re-tread all of the existing stuff 😆

**Pull request merged**